core/src/main/java/com/linecorp/armeria/common/util/TimeoutController.java

```diff
@@ -24,32 +26,79 @@
      */
-    void initTimeout(long timeoutMillis);
+    boolean scheduleTimeout(long timeoutMillis);
 
     /**
      * Extends the current timeout by the specified {@code adjustmentMillis}.
      * Note that a negative {@code adjustmentMillis} reduces the current timeout.
      *
      * @param adjustmentMillis the adjustment of time amount value in milliseconds.
+     * @return {@code true} if the current timeout is extended by the specified {@code adjustmentMillis}.
+     *         {@code false} if no timeout was scheduled previously, the timeout has been triggered already
+     *         or a timeout cannot be scheduled, e.g. request or response has been handled already.
      */
-    void extendTimeout(long adjustmentMillis);
+    boolean extendTimeout(long adjustmentMillis);
 
     /**
      * Sets the amount of time that is after the specified {@code newTimeoutMillis} from now.
      *
      * @param newTimeoutMillis the new timeout value in milliseconds. {@code 0} if disabled.
+     * @return {@code true} if the current timeout is reset by the specified {@code newTimeoutMillis}.
+     *         {@code false} if the timeout has been triggered already
+     *         or a timeout cannot be scheduled, e.g. request or response has been handled already.
      */
-    void resetTimeout(long newTimeoutMillis);
+    boolean resetTimeout(long newTimeoutMillis);
 
     /**
-     * Cancels the current timeout scheduled.
+     * Trigger the current timeout immediately.
+     *
+     * @return {@code true} if the current timeout is triggered successfully.
+     *         {@code false} if no timeout was scheduled previously, the timeout has been triggered already
+     *         or a timeout cannot be scheduled, e.g. request or response has been handled already.
+     */
+    boolean timeoutNow();
+
+    /**
+     * Cancels the current timeout scheduled. You can schedule a new timeout with
+     * {@link #scheduleTimeout(long)} if the current timeout is canceled successfully.
+     * @return {@code true} if the current timeout is canceled.
+     *         {@code false} if the timeout has been triggered already or no timeout was scheduled previously.
      */
     boolean cancelTimeout();
 
     /**
-     * Returns the start time of the initial timeout in nanoseconds.
+     * Returns whether the timeout is scheduled or not.
+     *
+     * @return {@code true} the timeout is scheduled now.
+     *         {@code false} if the timeout has been triggered already or no timeout was scheduled previously.
+     */
+    boolean isScheduled();
+
+    /**
+     * Returns whether the timeout has been triggered or not.
+     *
+     * @return {@code true} if the timeout has been triggered already.
+     *         {@code false} if the timeout is scheduled now or no timeout was scheduled previously.
+     */
+    boolean isTimedOut();
+
+    /**
+     * Returns {@code true} if no timeout was scheduled previously.
+     */
+    boolean isDisabled();
```

Comment:
Seems like this is only used for testing.

I'm not sure, but if (`startTimeNanos()` is `null` if and only if `state` is `DISABLED`), then we could:

- Remove `isDisabled()` and use `startTimeNanos() == null` in the tests
- Replace `isExecutedAtLeastOnce` with `state != DISABLED`
  - We could set `firstExecutionTime` when the current state is `DISABLED`.

Reply:
The `DISABLED` state can be reached even if `isExecutedAtLeastOnce = true` by calling `cancelTimeout()`
```java
TimeoutContoller tc = new DefaultTimeoutController() // isExecutedAtLeastOnce == false, DISABLED
tc.scheduleTimeout(1000) // isExecutedAtLeastOnce == true, SCHEDULED
// cancel timeout before the deadline.
tc.cancelTimeout() // isExecutedAtLeastOnce == true, DISABLED
```

By the way, I think we can make `startTimeNanos() == null` after calling `cancelTimeout()`.
It is a nicer approach.

Reply:
The `responseTimeoutMillis` in `DefaultClientContext` is used to represent the amount of time since the start time of receiving the response.

```java
ctx.setResponseTimeoutMillis(1000)
// 1 seconds after
ctx.setResponseTimeoutAfterMillis(1000) 
// We need startTimeNanos() to get new 'responseTimeoutMillis' 
assert ctx.responseTimeoutMillis == System.nanoTime() - statTimeNanos() + responseTimeoutAfterMillis
```

I think we still need `isExecutedAtLeastOnce()` or add a new state like `WAITING` to express not never scheduled?

Reply:
+1 for adding another state. Probably `INIT`?

Reply:
Good, thought `INIT` too. 👍